chain/src/main/scala/org/bitcoins/chain/blockchain/sync/FilterSync.scala

```diff
@@ -0,0 +1,139 @@
+      logger.info(
+        s"Our filters are synced with our peers filters, both at blockHash=${ourBestFilterHeader.blockHashBE}")
+      Future.successful(chainApi)
+    } else {
+      logger.info(
+        s"Beginning sync for filters from filterheader=${ourBestFilterHeader} to blockheader=${ourBestHeader.hashBE}")
+      //let's fetch all missing filter headers first
+      val bestFilterBlockHeaderF =
+        chainApi.getHeader(ourBestFilterHeader.blockHashBE)
+
+      val headersMissingFiltersF = for {
+        bestFilterBlockHeader <- bestFilterBlockHeaderF
+        missing <- chainApi.getHeadersBetween(from = bestFilterBlockHeader.get,
+                                              to = ourBestHeader)
+      } yield missing
+
+      //now that we have headers that are missing filters, let's fetch the filters
+      val fetchFiltersF: Future[Vector[(BlockHeaderDb, GolombFilter)]] = for {
+        missingHeaders <- headersMissingFiltersF
+        fetchNested = missingHeaders.map { b =>
+          val filterF = getFilterFunc(b.blockHeader)
+          filterF.map(f => (b, f))
+        }
+        fetched <- Future.sequence(fetchNested)
+      } yield fetched
+
+      //now let's build filter headers
+      val blockFiltersAggF: Future[Vector[BlockFilterAggregated]] = {
+        fetchFiltersF.map {
+          case filters: Vector[(BlockHeaderDb, GolombFilter)] =>
+            buildBlockFilterAggregated(filters, ourBestFilterHeader)
+        }
+      }
+
+      val compactFiltersF = blockFiltersAggF.map { filtersAgg =>
+        filtersAgg.map { agg =>
+          CompactFilterMessage(blockHash = agg.blockHeader.hash,
+                               filter = agg.filter)
+        }
+      }
+
+      val blockHeaderF = blockFiltersAggF.map { filtersAgg =>
+        filtersAgg.last.blockHeader
+      }
+      val filterHeadersF = blockFiltersAggF.map(_.map(_.filterHeader))
+      for {
+        blockHeader <- blockHeaderF
+        compactFilters <- compactFiltersF
+        filterHeaders <- filterHeadersF
+        headersChainApi <- chainApi.processFilterHeaders(filterHeaders,
+                                                         blockHeader.hashBE)
+        filtersChainApi <- headersChainApi.processFilters(compactFilters)
+      } yield filtersChainApi
+    }
+  }
+
+  private def buildBlockFilterAggregated(
+      filters: Vector[(BlockHeaderDb, GolombFilter)],
+      ourBestFilterHeader: CompactFilterHeaderDb): Vector[
+    BlockFilterAggregated] = {
+    filters.zipWithIndex.map {
+      case ((blockHeaderDb, filter), idx) =>
+        val filterHeader = if (idx == 0) {
+          FilterHeader(filterHash = filter.hash,
+                       prevHeaderHash = ourBestFilterHeader.hashBE.flip)
+        } else {
+          val prev = filters(idx - 1)._2.hash
+          FilterHeader(filterHash = filter.hash, prevHeaderHash = prev)
+        }
+        BlockFilterAggregated(filterHeader, filter, blockHeaderDb.blockHeader)
```

Comment:
I think it's worth mentioning that this class can be only used with a trusted Bitcoin Core node, because it doesn't verify filters against filter header, but rather generates filter headers by its own.